projects/power_distribution/test/test_fan_ctrl.c

```diff
@@ -0,0 +1,219 @@
+  .i2c = TEST_I2C_PORT,
+  .fan_pwm1 = TEST_PWM_1,
+  .fan_pwm2 = TEST_PWM_2,
+};
+
+static uint16_t adc_ret_val;
+StatusCode TEST_MOCK(adc_read_converted_pin)(GpioAddress address, uint16_t *reading) {
+  *reading = adc_ret_val;
+  return STATUS_CODE_OK;
+}
+
+static uint8_t i2c_buf1[10];
+static uint8_t i2c_buf2[10];
+StatusCode TEST_MOCK(i2c_write)(I2CPort i2c, I2CAddress addr, uint8_t *tx_data, size_t tx_len) {
+  // Save both temp values written over i2c
+  if (tx_data[0] == ADT7476A_PWM_1) {
+    for (size_t i = 0; i < tx_len; i++) {
+      i2c_buf1[i] = tx_data[i];
+    }
+  } else {
+    for (size_t i = 0; i < tx_len; i++) {
+      i2c_buf2[i] = tx_data[i];
+    }
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode TEST_MOCK(i2c_read_reg)(I2CPort i2c, I2CAddress addr, uint8_t reg, uint8_t *rx_data,
+                                   size_t rx_len) {
+  if (reg == ADT7476A_INTERRUPT_STATUS_REGISTER_1) {
+    *rx_data = OVERVOLT_FLAGS;
+  } else if (reg == ADT7476A_INTERRUPT_STATUS_REGISTER_2) {
+    *rx_data = FAN_ERR_FLAGS;
+  }
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_can_fan_ctrl_rx_handler(const CanMessage *msg, void *context,
+                                              CanAckStatus *ack_reply) {
+  LOG_DEBUG("\nFAN CTRL RX CALLBACK ACTIVATED\n");
+  s_fan_ctrl_msg = msg->data;
+  return STATUS_CODE_OK;
+}
+
+static void prv_initialize_can(SystemCanDevice can_device) {
+  CanSettings can_settings = {
+    .device_id = can_device,
+    .loopback = true,
+    .bitrate = CAN_HW_BITRATE_500KBPS,
+    .rx_event = TEST_CAN_EVENT_RX,
+    .tx_event = TEST_CAN_EVENT_TX,
+    .fault_event = TEST_CAN_EVENT_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+  };
+  can_init(&s_can_storage, &can_settings);
+}
+
+void setup_test(void) {
+  gpio_init();
+  interrupt_init();
+  soft_timer_init();
+  event_queue_init();
+  adc_init(ADC_MODE_SINGLE);
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,
+    .scl = TEST_CONFIG_PIN_I2C_SCL,
+    .sda = TEST_CONFIG_PIN_I2C_SDA,
+  };
+  i2c_init(TEST_I2C_PORT, &i2c_settings);
+  gpio_it_init();
+}
+
+void teardown_test(void) {}
+
+void test_front_pd_fan_ctrl_init(void) {
+  TEST_ASSERT_NOT_OK(pd_fan_ctrl_init(NULL, true));
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, true));
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, false));
+}
+
+// Test Gpio interrupt on rear smbalert pin triggered
+void test_fan_err_rear(void) {
+  prv_initialize_can(SYSTEM_CAN_DEVICE_POWER_DISTRIBUTION_REAR);
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, false));
+  gpio_it_trigger_interrupt(&(GpioAddress)REAR_PIN_SMBALERT);
+  can_register_rx_handler(SYSTEM_CAN_MESSAGE_REAR_FAN_FAULT, prv_can_fan_ctrl_rx_handler, NULL);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg >> 8 & OVERVOLT_FLAGS) == OVERVOLT_FLAGS);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg & FAN_ERR_FLAGS) == FAN_ERR_FLAGS);
+  s_fan_ctrl_msg = 0;
```

Comment:
Instead of resetting this at the end of every test, you can put it in `setup_test` or `teardown_test`.